### ISSUE TITLE
Better vendor / asset path finding.

### ISSUE THREAD

**Issue by SamMousa** [yiisoft/yii2#10605]
Currently Yii2 offers no way to automatically "find" the vendor path.

Basically there are 2 options:
1. Put it in the default location.
2. Put it somewhere else and set the path in the configuration.

I think I have found an alternative (https://github.com/francoispluchino/composer-asset-plugin/issues/183):

````
$vendorPath = dirname(dirname((new \ReflectionClass(\Composer\Autoload\ClassLoader::class))->getFileName()))
````

This way we can always find the vendor path and when we reconfigure composer.json we do not have to reconfigure the application.

We could even go one step further and set the aliases `@bower` and `@npm` more intelligently by just reading the `installed.json` file in the composer directory. (This is what is done in the Symfony bundle https://github.com/francoispluchino/FxpRequireAssetBundle)

Pros:
1. Less configuration
2. More robust

Con:
1. Theoretically a little less performant since installed.json gets read on each request.

-- Note, I did some tests on a real installed.json file for a yii2 project, size about 160KB, running on core i2500k inside a homestead vm, got the following results doing 20 iterations of 1000 loads:
````
Average 1000x: 1.6277709126472
Max 1000x: 1.7626240253448
Min 1000x: 1.5363171100616
````
So basically this would add about `2ms` to page load times.

**Comment by francoispluchino** [comment]
@SamMousa For Symfony, it's the best practice, because the configuration is cached, and therefore, the analysis is not performed each time.

**Comment by francoispluchino** [comment]
As I indicated in francoispluchino/composer-asset-plugin#183, I will create a small library for performing this task, with the ability to cache the analysis.

So, @qiangxue, @cebe, @samdark if you're interested ...

**Comment by SamMousa** [comment]
Ah, that's nice, well in Yii that also woudn't be too difficult to do, although I do not think it's standard.
On the other hand Yii's configuration is PHP so it gets cached if APC is in use.

Anyway `yii2-team` let me know if you need anything from me regarding this, or if you will just follow up directly when @francoispluchino Creates his library ;-)

**Comment by samdark** [comment]
I don't like the idea of reflecting and reading from JSON on each request and don't really like to complicate config with additional caching which could result in debug hell easily.

**Comment by SamMousa** [comment]
Well reflecting is something Yii does for every request regardless; for example when calling controller actions, so that argument seems unfair to me.

Also reading from JSON is actually not that bad.
This is my configuration `ls` for a Yii2 project config directory:
````
4.0K	common.php
4.0K	console.php
0	debug
4.0K	debug.php
4.0K	dependencies.php
4.0K	env.php
20K	envs
4.0K	local.php
4.0K	web.php
````

It has some optional configuration files depending on environment and stuff like that. I ran  the same test as I did for the JSON file of 160KB.
````
Average 1000x: 2.8695501327515
Max 1000x: 3.0958609580994
Min 1000x: 2.7461588382721
````

So actually configuration in JSON is faster than (structured) configuration in php. Running this for just the `common.php` file beats JSON though:
````
Average 1000x: 0.55152471065521
Max 1000x: 0.6046290397644
Min 1000x: 0.52181482315063
````

That said these differences all seem minor and irrelevent to me.

Furthermore, we could execute the detection code only in a few cases where there is a guaranteed error waiting to happen anyway:
1. There is no vendor path in configuration.
2. The base path of the application does not have a vendor subdirectory.

In the current situation, the following alias would be set:
````
$this->setVendorPath($this->getBasePath() . DIRECTORY_SEPARATOR . 'vendor');
````
Regardless if the directory exists. If it does not that is just waiting for an exception as soon as you use it. 

I think usability is more important than this minor performance hit which can be removed by manual configuration regardless.

**Comment by dynasource** [comment]
@sammousa: you are actually giving the answer not to do it:
- it adds 2ms.

If it would take down 2ms, you have my vote.

To elaborate further, if I could save many of these '2ms' scenario's, just by doing some manual configuration, it would do this right away.
